### ISSUE TITLE
on() not triggered when using a mutable store item as a dependency

### ISSUE THREAD

**Issue by chvelkov** [solidjs/solid#1833]
### Describe the bug

Here's an example:

### Your Example Website or App

https://playground.solidjs.com/anonymous/64e29718-38db-4df4-afeb-4445b799d170

### Steps to Reproduce the Bug or Issue

Press Increment

### Expected behavior

When pressing the button I'd expect that both counters update, however the one using on() isn't working.

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by ryansolid** [comment]
Yeah this is akin to setting a value at an index in createStore. https://playground.solidjs.com/anonymous/6ea855b3-1f2d-4dc6-a9c0-e413b6258e64

It notifies things that listen to the index but not everything. The thing is that something that would change the length probably should trigger the parent.. we use length actually as a trick for this as iteration depends on it which means it works in all cases other than `on` which doesn't iterate. And since on is a blanket reactive wrapper (ie can return anything not just the thing that is accessed) there isn't really a way to address this on the `on` side.

The fact that `produce` is also impacted suggests this is something we should consider changing if feasible. Thanks for bringing this to my attention. I never use `on` or `mutable` interfaces so this wasn't in my direct thinking.

**Comment by ryansolid** [comment]
Yeah the solution here is something like deep. We've needed to solve this in 2.0 and there is a deep helper in Solid primitives. But this is just how `on` works and I don't see an feasible way to resolve this.
